### ISSUE TITLE
Note sometimes gets 'stuck' when using portamento

### ISSUE THREAD

**Issue by DieterVDW** [DieterVDW/arduino-usb-midi-to-CV#1]
Pretty easy to reproduce.

**Comment by nshaver** [comment]
I finally got my littlebit synth running again, but I'm not exactly using this code. I'm using a modified version that doesn't require arcore for its MIDI, and it uses a dedicated serial MIDI module that I built using a Perf bit. I don't foresee using my USB MIDI (arcore) version anymore. It's just so much simpler having the dedicated MIDI, no USB hub required, no iPad required.

Nevertheless, I was able to reproduce this issue by turning up the portamento just a bit, playing a note and then quickly playing another note that's several notes higher, and then quickly releasing both notes at the same time. For instance, an A2 and then an A3, and quickly release both notes before the portamento can glide all the way up to the A3=problem.

If I play an A2 and then a B2 the issue doesn't occur. The glide gets up to the B2 so fast I can't create the issue.

If I play an A2 and then an A3, but I wait for the glide to get to A3 - problem occurs.

If I play an A2 and then an A3, and then release the A3, then release A2 - no problem.

So, I think it has something to do with not detecting the release when it hasn't yet reached the note that it's gliding to.

I'm going to work on it some tonight, see if I can find exactly where it's happening, and fix it. Once I figure it out, I'll checkin a version here. I may also try to checkin a version that supports both arcore USBMIDI and serial MIDI.

**Comment by DieterVDW** [comment]
Awesome! Don't feel pressured though, it's only for hobby and fun right?

Nice thing would also be to make the portamento actually glide ...
I think now it just plays the notes in between?
What do you think, think that would be possible?

I have the littleBits MIDI bit, but this is actually a lot better!
Portamento, modulation, pitchwheel! Superb!

**Comment by nshaver** [comment]
Yeah, that portamento does step if you set it too slow. If it's set pretty low though, it seems to glide very naturally. I bet I can figure that out.

I had that MIDI bit too, and was very disappointed in it. I can't see how anybody would enjoy it very much without portamento, pitch wheel, and modulation. I wound up sending mine back in return for the Perf bit that I used to make my serial MIDI interface.

**Comment by DieterVDW** [comment]
Yeah indeed, your Arduino solution is way better!
And works with my Mininova over USB (after the NOTE OFF fix), so pretty ideal!

Unfortunatly the littleBits are pretty hard (and expensive!) to come by here in Europe, and "return if not satisfied" doesn't really apply here I'm afraid ... So yeah, think I'm stuck with a pretty useless MIDI bit :) .

(BTW: no problems with tuning when using live?)

**Comment by nshaver** [comment]
Tuning is definitely an issue. You can easily get each oscillator tuned to a single note, but the biggest problem is that they don't stay equally in tune as you go up and down the keyboard. I have solved that by putting a "dimmer" bit in front of one oscillator. With that dimmer bit fully to the right (zero resistance), it has no effect. But if you turn it down slightly, it slightly lowers the voltage that goes into that oscillator, and that serves as a "scale" adjuster, and I can usually get it set where both oscillators are equally tuned up and down the keyboard.

**Comment by nshaver** [comment]
I think I've got the stuck note issue fixed. I tried to upload my change but I don't think I've got access to this project. Can you add me as a collaborator - nick.shaver@gmail.com is my username.

**Comment by DieterVDW** [comment]
I added the username 'nshaver' . Github said it didn't know a user with that email address ... ?

**Comment by nshaver** [comment]
Thanks. That was me. It's checked in now, and I think it's fixed. That change fixed my version (serial MIDI, not USB MIDI).

**Comment by DieterVDW** [comment]
Alright! I'll try it out first chance I get (won't be in the first few days though..) .

**Comment by nshaver** [comment]
I've been working on the issue with the portamento "stepping" when it's set to a high (slow) value. I thought surely I could fix that, because when set to a fast value it doesn't seem to have that problem. After researching things a little more though, I don't think it's fixable. It actually does the same thing when using a fast portamento, you just don't notice it because it moves so fast through the steps. Here's the deal - we only have 255 different values that can be sent out via analogWrite, and those 255 values have to cover the entire frequency range of the synth. The issue might be able to be solved by manually implementing your own PWM frequency, and doing something like this rather than simply letting analogWrite(x) do it for you:

digitalWrite(NOTE_PIN1, HIGH);
delayMicroseconds(x);
digitalWrite(NOTE_PIN1, LOW);
delayMicroseconds(freq-x);

I think with manual manipulation of the PWM timing, a lot more than 255 different frequencies could be reached, thus solving the "stepping" issue. That would come with timing issues though because rather than using the Arduino's built-in hardware PWM timer, the code would have to constantly execute those four lines of code in order to keep the PWM wave going. It sounds really really hard to me when compared to just letting the PWM timer do its thing.

So, on that other portamento stepping issue - I don't think it's worth the timing issues that would result from trying to solve it.
